### ISSUE TITLE
most of time for backup is cost in scan(unreasonable)

### ISSUE THREAD

**Issue by gluthgao** [restic/restic#1160]
## Output of `restic version`

debug log file /Users/gluth/go/src/github.com/restic/restic/log1.txt
debug enabled
restic 0.7.1 
compiled with go1.8.1 on darwin/amd64
## How did you run restic exactly?
Run the back up of local NAS to remote NAS
./restic -r remoteNAS backup /  NAS

## What backend/server/service did you use?
NAS

## Expected behavior
We should not scan the file system first which cost long time, i have lots of file  around 15TB, 
the total time should be spend on the backup data.
## Actual behavior
the scan cost around 6 hours !!!!

## Steps to reproduce the behavior
lots of file system like NAS, then do the backup

## Do you have any idea what may have caused this?
scan file system cost more time is normally since the fs.walk is single thread travel model which will cost long time.

## Do you have an idea how to solve the issue?
Do we necessary to scan the file system to get the total file/dir number + size?
also some file may not have permission to read or read failure that's hard to get the accurate size.

**Comment by fd0** [comment]
Hi, thanks for raising this issue. First, please let me give you feedback on your style of reporting issues here at GitHub: I feel that you're writing in a style that comes across as very accusatory and heated (at least to me). Can you try writing more matter-of-factly? That'd help discussing issues a lot. Thanks!

So what I understood so far is that you have a local directory called `NAS` with about 15TiB of data that is to be saved into a local repository in the directory `remoteNAS`, is that correct? From the names of the directories I can guess that the data is really read from (and saved to) network attached storage, so how did you mount the directories? Which service and implementation is used for that?

Which type of storage is used for both the directory save and the storage location for the repository? Classical hard discs or flash-based SSDs?

The scan phase is single-threaded because in my benchmarks having a multi-thread scanning phase did not perform any better, but instead slowed finding all file and directory sizes down.

We're scanning the file system to be saved before a backup for two reasons:
 * To give the user a sense of where we are in the backup it is necessary to find the number of files and the sum of the file sizes
  * This way, on modern operating systems like Linux the directory cache is primed. The backup phase that follows right after scanning can then be much more efficient because restic and the kernel doesn't need to wait until a directory listing is loaded from disc.

**Comment by gluthgao** [comment]
I‘m not good at write english:), you guys did a great job,  i'm in looking to use restic to backup NAS.
The environment is i mounted a NAS system with high speed network connection, so there is no performance issue in the network, my point is i compared to other backup software like commvault/avamar/veeam there is no scan first to try to get the total dirs/files, this step in my env have 1000000000 files which cost long time to scan, they just do the backup, when finished just gave a summary info like how much data backcuped to backup server.  thanks

**Comment by phillipp** [comment]
Hi,
I wanted to add my 2 cents here. Thinking about these two reasons, they don't make sense in all backup situations:

- In an automated backup (which is IMO the desired case), scanning the contents before making the backups is unnecessary work, because there is no progress to be displayed
- In a non-snapshotted filesystem, there may be inconsistencies when the user and/or applications removes/adds files and/or directories between scan and backup and the progress is not correct 
- In larger production setups (like ours) the filesystem cache will never fit into RAM. When the files are pruned from the cache, the backup phase will re-read *all* the directory entries from disk again, so every dentry is read twice instead of once

In our production setup (where backup is non-interactive and the directory entry cache will never fit into RAM, but only a small portion of files files change regularly) skipping the scan phase should theoretically cut the backup time into half.

Is there any possibility that the scan could be disabled via switch/flag, or is the scan a necessity?

**Comment by fd0** [comment]
For the record: I agree to your first and third point, and we can make restic more intelligent to skip scanning in these cases.

There's no option (yet) to disable scanning.

**Comment by phillipp** [comment]
I played with the code and just commented out the scan phase and gave the archiver an empty progress, so skipping the scan phase is very simple, code-wise.

I would like to implement this, but wanted to get some feedback on this first. IMHO the best idea is a new switch --no-progress (which is somewhat contrary to the unix philosophy of "no progress unless requested", but for a backup program this might be okay).

I was thinking about using the --quiet flag for this, but at least in our use-case we need to take a backup (without scanning and progress) and still need the new snapshot ID for adding it to our backup and restoe automation system.

So a *new option --no-progress* seems like the best way to go - **or**, but this is somewhat more radical - scanning and *progress is only enabled by a --progress flag*, which is more like all the rest of unix tools does it (rsync, for example).

@fd0 what do you think?

**Comment by fd0** [comment]
Let me think about this for a bit. Before we do anything in this matter, there's also an open PR (#1095) that modifies the status printing code (and I still need to review it). So hold on a little longer ;)

**Comment by remss** [comment]
The progress feedback is nice, the problem is that nothing is copied until listing files is done.
Can't we just run the initial scan in the background and let the display progress use the last total_size / total_nbfiles values?

**Comment by wscott** [comment]
It would be nice the actual backup was started in another thread when the first modified file was noticed by the initial scan. Then the progress bar could have both the finish and total numbers updating at the same time as until the scan has completed. This still keeps the caching advantages as the scanner would be ahead of the backup stream. I also suspect it would feel better interactively because you wouldn't have the next pause before the real backup progress bar starts to be displayed.

Sorry I haven't read the code so I don't know how much work my idea would cause. 

Having a --no-progress option for backups run from cron (the normal case) would also be welcome.

**Comment by mika-fischer** [comment]
Just to add my two cents: This issue is the reason why we can't use restic to backup our >10TB NAS with many many small files. The scan takes several hours...

An option to disable the progress display would be really appreciated.

Great work otherwise, I really would like to be able to use restic! :-)

**Comment by phillipp** [comment]
@mika-fischer 

Just apply the following patch and you're good.

```
diff --git a/cmd/restic/cmd_backup.go b/cmd/restic/cmd_backup.go
index 67378e1..bc5f20d 100644
--- a/cmd/restic/cmd_backup.go
+++ b/cmd/restic/cmd_backup.go
@@ -451,10 +451,12 @@ func runBackup(opts BackupOptions, gopts GlobalOptions, args []string) error {
                return true
        }
 
-       stat, err := archiver.Scan(target, selectFilter, newScanProgress(gopts))
-       if err != nil {
-               return err
-       }
+       // stat, err := archiver.Scan(target, selectFilter, newScanProgress(gopts))
+       // if err != nil {
+       //      return err
+       // }
+
+       stat := restic.Stat{}
 
        arch := archiver.New(repo)
        arch.Excludes = opts.Excludes

```

**Comment by fd0** [comment]
FYI, this has partly been addressed in #1676 (scan isn't done when `--quiet` is specified) and will be resolved by the new archiver code in #1494: The scan is run concurrently to the backup itself.

**Comment by bachp** [comment]
Just to give some feedback. Adding the `--quiet` brought down the incremental backup time from over 8 hours down to 12 minutes.

**Comment by mholt** [comment]
@bachp Using which version of restic?

**Comment by fd0** [comment]
I think we can close this issue, it is resolved in restic >= 0.9.0. Please feel free to add further comments. Thanks!
